Linker failure: undefined reference to `z_deflateInit_'

**Operating system or device - Godot version:**

GNU/Linux, Parabola or Arch Linux, is almost the same thing.

**Issue description** (what happened, and what was expected):

Compiling Godot from sources, I'm the AUR package maintainer for the engine. This is the output I get from the linking phase:

```
core/libcore.x11.opt.tools.32.llvm.a(compression.x11.opt.tools.32.llvm.o): In function `Compression::compress(unsigned char*, unsigned char const*, int, Compression::Mode)':
core/io/compression.cpp:(.text+0x94): undefined reference to `z_deflateInit_'
core/io/compression.cpp:(.text+0xae): undefined reference to `z_deflateBound'
core/io/compression.cpp:(.text+0xcd): undefined reference to `z_deflate'
core/io/compression.cpp:(.text+0xdd): undefined reference to `z_deflateEnd'
core/libcore.x11.opt.tools.32.llvm.a(compression.x11.opt.tools.32.llvm.o): In function `Compression::get_max_compressed_buffer_size(int, Compression::Mode)':
core/io/compression.cpp:(.text+0x1f4): undefined reference to `z_deflateInit_'
core/io/compression.cpp:(.text+0x20a): undefined reference to `z_deflateBound'
core/io/compression.cpp:(.text+0x218): undefined reference to `z_deflateEnd'
core/libcore.x11.opt.tools.32.llvm.a(compression.x11.opt.tools.32.llvm.o): In function `Compression::decompress(unsigned char*, int, unsigned char const*, int, Compression::Mode)':
core/io/compression.cpp:(.text+0x2ef): undefined reference to `z_inflateInit_'
core/io/compression.cpp:(.text+0x37e): undefined reference to `z_inflate'
core/io/compression.cpp:(.text+0x38c): undefined reference to `z_inflateEnd'
core/libcore.x11.opt.tools.32.llvm.a(zip.x11.opt.tools.32.llvm.o): In function `zipOpenNewFileInZip4_64':
core/io/zip.c:(.text+0x1dc4): undefined reference to `z_deflateInit2_'
core/io/zip.c:(.text+0x1e05): undefined reference to `z_get_crc_table'
core/libcore.x11.opt.tools.32.llvm.a(zip.x11.opt.tools.32.llvm.o): In function `zipWriteInFileInZip':
core/io/zip.c:(.text+0x2466): undefined reference to `z_crc32'
core/io/zip.c:(.text+0x2529): undefined reference to `z_deflate'
core/libcore.x11.opt.tools.32.llvm.a(zip.x11.opt.tools.32.llvm.o): In function `zipCloseFileInZipRaw64':
core/io/zip.c:(.text+0x2711): undefined reference to `z_deflate'
core/io/zip.c:(.text+0x2765): undefined reference to `z_deflateEnd'
core/libcore.x11.opt.tools.32.llvm.a(unzip.x11.opt.tools.32.llvm.o): In function `unzClose':
core/io/unzip.c:(.text+0xdbd): undefined reference to `z_inflateEnd'
core/libcore.x11.opt.tools.32.llvm.a(unzip.x11.opt.tools.32.llvm.o): In function `unzCloseCurrentFile':
core/io/unzip.c:(.text+0xe61): undefined reference to `z_inflateEnd'
core/libcore.x11.opt.tools.32.llvm.a(unzip.x11.opt.tools.32.llvm.o): In function `unzOpenCurrentFile3':
core/io/unzip.c:(.text+0x1fe0): undefined reference to `z_inflateEnd'
core/io/unzip.c:(.text+0x2647): undefined reference to `z_inflateInit2_'
core/libcore.x11.opt.tools.32.llvm.a(unzip.x11.opt.tools.32.llvm.o): In function `unzSeekCurrentFile':
core/io/unzip.c:(.text+0x27c3): undefined reference to `z_inflateReset'
core/libcore.x11.opt.tools.32.llvm.a(unzip.x11.opt.tools.32.llvm.o): In function `unzReadCurrentFile':
core/io/unzip.c:(.text+0x29ee): undefined reference to `z_crc32'
core/io/unzip.c:(.text+0x2a51): undefined reference to `z_inflate'
core/io/unzip.c:(.text+0x2a96): undefined reference to `z_crc32'
clang-3.9: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [bin/godot.x11.opt.tools.32.llvm] Error 1
scons: building terminated because of errors.
==> ERROR: Se produjo un fallo en build().
    Cancelando…
```

**Steps to reproduce:**

Try my package at AUR on Arch Linux and you will see.

**Link to minimal example project** (optional but very welcome):

--- a/core/SCsub
+++ b/core/SCsub
@@ -99,6 +99,8 @@ if env["builtin_zlib"]:
     env.Prepend(CPPPATH=[thirdparty_zlib_dir])
     if env.dev_build:
         env_thirdparty.Append(CPPDEFINES=["ZLIB_DEBUG"])
+        # Affects headers so it should also be defined for Godot code
+        env.Append(CPPDEFINES=["ZLIB_DEBUG"])
 
     env_thirdparty.add_source_files(thirdparty_obj, thirdparty_zlib_sources)
 

--- a/modules/freetype/SCsub
+++ b/modules/freetype/SCsub
@@ -67,8 +67,6 @@ if env["builtin_freetype"]:
     env.Prepend(CPPPATH=[thirdparty_dir + "/include"])
 
     env_freetype.Append(CPPDEFINES=["FT2_BUILD_LIBRARY", "FT_CONFIG_OPTION_USE_PNG", "FT_CONFIG_OPTION_SYSTEM_ZLIB"])
-    if env.dev_build:
-        env_freetype.Append(CPPDEFINES=["ZLIB_DEBUG"])
 
     # Also requires libpng headers
     if env["builtin_libpng"]:

--- a/platform/linuxbsd/detect.py
+++ b/platform/linuxbsd/detect.py
@@ -222,23 +222,6 @@ def configure(env: "SConsEnvironment"):
 
     # FIXME: Check for existence of the libs before parsing their flags with pkg-config
 
-    # freetype depends on libpng and zlib, so bundling one of them while keeping others
-    # as shared libraries leads to weird issues. And graphite and harfbuzz need freetype.
-    ft_linked_deps = [
-        env["builtin_freetype"],
-        env["builtin_libpng"],
-        env["builtin_zlib"],
-        env["builtin_graphite"],
-        env["builtin_harfbuzz"],
-    ]
-    if (not all(ft_linked_deps)) and any(ft_linked_deps):  # All or nothing.
-        print_error(
-            "These libraries should be either all builtin, or all system provided:\n"
-            "freetype, libpng, zlib, graphite, harfbuzz.\n"
-            "Please specify `builtin_<name>=no` for all of them, or none."
-        )
-        sys.exit(255)
-
     if not env["builtin_freetype"]:
         env.ParseConfig("pkg-config freetype2 --cflags --libs")
 
@@ -250,6 +233,11 @@ def configure(env: "SConsEnvironment"):
 
     if not env["builtin_harfbuzz"]:
         env.ParseConfig("pkg-config harfbuzz harfbuzz-icu --cflags --libs")
+
+    if not env["builtin_icu4c"] or not env["builtin_harfbuzz"]:
+        print_warning(
+            "System-provided icu4c or harfbuzz cause known issues for GDExtension (see GH-91401 and GH-100301)."
+        )
 
     if not env["builtin_libpng"]:
         env.ParseConfig("pkg-config libpng16 --cflags --libs")

